cli/src/commands/revert.rs

```diff
@@ -0,0 +1,180 @@
+    let mut workspace_command = command.workspace_helper(ui)?;
+    let to_revert: Vec<_> = workspace_command
+        .parse_union_revsets(ui, &args.revisions)?
+        .evaluate_to_commits()?
+        .try_collect()?; // in reverse topological order
+    if to_revert.is_empty() {
+        writeln!(ui.status(), "No revisions to revert.")?;
+        return Ok(());
+    }
+    let (new_parents, new_children) = compute_commit_location(
+        ui,
+        &workspace_command,
+        args.destination.as_deref(),
+        args.insert_after.as_deref(),
+        args.insert_before.as_deref(),
+        "reverted commits",
+    )?;
+    let transaction_description = if to_revert.len() == 1 {
+        format!("revert commit {}", to_revert[0].id().hex())
+    } else {
+        format!(
+            "revert commit {} and {} more",
+            to_revert[0].id().hex(),
+            to_revert.len() - 1
+        )
+    };
+    let commits_to_revert_with_new_commit_descriptions = {
+        let template_text = command
+            .settings()
+            .get_string("templates.revert_description")?;
+        let template = workspace_command.parse_commit_template(ui, &template_text)?;
+
+        to_revert
+            .into_iter()
+            .map(|commit| {
+                let mut output = Vec::new();
+                template
+                    .format(&commit, &mut PlainTextFormatter::new(&mut output))
+                    .expect("write() to vec backed formatter should never fail");
+                // Template output is usually UTF-8, but it can contain file content.
+                let commit_description = output.into_string_lossy();
+                (commit, commit_description)
+            })
+            .collect_vec()
+    };
+    let mut tx = workspace_command.start_transaction();
+    let mut parents = new_parents;
+    let original_parent_commit_ids: HashSet<_> = parents.iter().ids().cloned().collect();
+    let mut new_base_tree = merge_commit_trees(tx.repo(), &parents)?;
+
+    for (commit_to_revert, new_commit_description) in
+        &commits_to_revert_with_new_commit_descriptions
+    {
+        let old_base_tree = commit_to_revert.parent_tree(tx.repo())?;
+        let old_tree = commit_to_revert.tree()?;
+        let new_tree = new_base_tree.merge(&old_tree, &old_base_tree)?;
+        let new_parent_ids = parents.iter().map(|commit| commit.id().clone()).collect();
+        let new_commit = tx
+            .repo_mut()
+            .new_commit(new_parent_ids, new_tree.id())
+            .set_description(new_commit_description)
+            .write()?;
+        parents = vec![new_commit];
+        new_base_tree = new_tree;
+    }
+
+    let mut num_rebased = 0;
+    for child_commit in new_children {
+        let new_parent_ids = child_commit
+            .parent_ids()
+            .iter()
+            .filter(|id| !original_parent_commit_ids.contains(id))
+            .cloned()
+            .chain(parents.iter().ids().cloned())
```

Comment:
nit: maybe we should replace the first `original_parent_commit_ids` to `parents` to preserve the order? I think `duplicate` works that way.